From ebdd6b2d5e4151175aae2f07a069ffede9417295 Mon Sep 17 00:00:00 2001
From: nift4 <nift4@protonmail.com>
Date: Thu, 17 Aug 2023 19:34:38 +0200
Subject: [PATCH 1/5] Support A-GPS toggle

bind_nlp_setting_to_supl needs to be enabled in
/vendor/etc/agps_profiles_conf2.xml before this commit has any
effect. This works by providing a fake NLP setting status to mtk_agpsd,
which then uses it to en/disable A-GPS.
---
 .../location/lppe/main/LPPeService$13.smali   | 43 ++++++++++
 .../location/lppe/main/LPPeService.smali      | 78 +++++++++++++++----
 2 files changed, 106 insertions(+), 15 deletions(-)
 create mode 100644 smali/com/mediatek/location/lppe/main/LPPeService$13.smali

diff --git a/smali/com/mediatek/location/lppe/main/LPPeService$13.smali b/smali/com/mediatek/location/lppe/main/LPPeService$13.smali
new file mode 100644
index 0000000..3eee92a
--- /dev/null
+++ b/smali/com/mediatek/location/lppe/main/LPPeService$13.smali
@@ -0,0 +1,43 @@
+.class Lcom/mediatek/location/lppe/main/LPPeService$13;
+.super Landroid/database/ContentObserver;
+
+
+# annotations
+.annotation system Ldalvik/annotation/EnclosingClass;
+    value = Lcom/mediatek/location/lppe/main/LPPeService;
+.end annotation
+
+.annotation system Ldalvik/annotation/InnerClass;
+    accessFlags = 0x0
+    name = null
+.end annotation
+
+
+# instance fields
+.field final synthetic this$0:Lcom/mediatek/location/lppe/main/LPPeService;
+
+
+# direct methods
+.method constructor <init>(Lcom/mediatek/location/lppe/main/LPPeService;)V
+    .locals 1
+
+    iput-object p1, p0, Lcom/mediatek/location/lppe/main/LPPeService$13;->this$0:Lcom/mediatek/location/lppe/main/LPPeService;
+
+    const/4 v0, 0x0
+
+    invoke-direct {p0, v0}, Landroid/database/ContentObserver;-><init>(Landroid/os/Handler;)V
+
+    return-void
+.end method
+
+
+# virtual methods
+.method public onChange(Z)V
+    .locals 1
+
+    iget-object v0, p0, Lcom/mediatek/location/lppe/main/LPPeService$13;->this$0:Lcom/mediatek/location/lppe/main/LPPeService;
+
+    invoke-virtual {v0}, Lcom/mediatek/location/lppe/main/LPPeService;->lbsNlpStatusUpdate()V
+
+    return-void
+.end method
diff --git a/smali/com/mediatek/location/lppe/main/LPPeService.smali b/smali/com/mediatek/location/lppe/main/LPPeService.smali
index 1e1c565..02f8735 100644
--- a/smali/com/mediatek/location/lppe/main/LPPeService.smali
+++ b/smali/com/mediatek/location/lppe/main/LPPeService.smali
@@ -2249,11 +2249,6 @@
 
     invoke-virtual {v0, v2}, Landroid/content/IntentFilter;->addAction(Ljava/lang/String;)V
 
-    .line 371
-    const-string v2, "android.location.PROVIDERS_CHANGED"
-
-    invoke-virtual {v0, v2}, Landroid/content/IntentFilter;->addAction(Ljava/lang/String;)V
-
     .line 372
     const-string v2, "com.android.ims.IMS_SERVICE_UP"
 
@@ -2299,9 +2294,11 @@
     invoke-virtual {v0, v1}, Lcom/mediatek/location/lppe/ipaddr/LPPeIpAddress$LPPeIpAddressReceiver;->requestCapabilities(I)V
 
     .line 382
-    iget-object p0, p0, Lcom/mediatek/location/lppe/main/LPPeService;->mLbsReceiver:Lcom/mediatek/location/lppe/lbs/LPPeLbs$LPPeLbsReceiver;
+    iget-object v0, p0, Lcom/mediatek/location/lppe/main/LPPeService;->mLbsReceiver:Lcom/mediatek/location/lppe/lbs/LPPeLbs$LPPeLbsReceiver;
+
+    invoke-virtual {v0, v1}, Lcom/mediatek/location/lppe/lbs/LPPeLbs$LPPeLbsReceiver;->requestCapabilities(I)V
 
-    invoke-virtual {p0, v1}, Lcom/mediatek/location/lppe/lbs/LPPeLbs$LPPeLbsReceiver;->requestCapabilities(I)V
+    invoke-direct {p0}, Lcom/mediatek/location/lppe/main/LPPeService;->registerSettingsObserver()V
 
     .line 384
     const-string p0, "initService() done"
@@ -2942,20 +2939,71 @@
     .locals 0
 
     .line 1631
-    invoke-direct {p0}, Lcom/mediatek/location/lppe/main/LPPeService;->lbsNlpStatusUpdate()V
+    invoke-virtual {p0}, Lcom/mediatek/location/lppe/main/LPPeService;->lbsNlpStatusUpdate()V
 
     return-void
 .end method
 
-.method private lbsNlpStatusUpdate()V
-    .locals 3
+.method private registerSettingsObserver()V
+    .locals 7
 
-    .line 1622
-    iget-object v0, p0, Lcom/mediatek/location/lppe/main/LPPeService;->mLocationManager:Landroid/location/LocationManager;
+    iget-object v0, p0, Lcom/mediatek/location/lppe/main/LPPeService;->mContext:Landroid/content/Context;
 
-    const-string v1, "network"
+    invoke-virtual {v0}, Landroid/content/Context;->getContentResolver()Landroid/content/ContentResolver;
+
+    move-result-object v1
+
+    const-string v2, "assisted_gps_enabled"
 
-    invoke-virtual {v0, v1}, Landroid/location/LocationManager;->isProviderEnabled(Ljava/lang/String;)Z
+    invoke-static {v2}, Landroid/provider/Settings$Global;->getUriFor(Ljava/lang/String;)Landroid/net/Uri;
+
+    move-result-object v3
+
+    const/4 v4, 0x0
+
+    new-instance v5, Lcom/mediatek/location/lppe/main/LPPeService$13;
+
+    invoke-direct {v5, p0}, Lcom/mediatek/location/lppe/main/LPPeService$13;-><init>(Lcom/mediatek/location/lppe/main/LPPeService;)V
+
+    sget v6, Landroid/os/UserHandle;->USER_ALL:I
+
+    invoke-virtual {v1, v3, v4, v5, v6}, Landroid/content/ContentResolver;->registerContentObserver(Landroid/net/Uri;ZLandroid/database/ContentObserver;I)V
+
+    return-void
+.end method
+
+.method private isAssistedGpsEnabled()Z
+    .locals 2
+
+    iget-object v0, p0, Lcom/mediatek/location/lppe/main/LPPeService;->mContext:Landroid/content/Context;
+
+    invoke-virtual {v0}, Landroid/content/Context;->getContentResolver()Landroid/content/ContentResolver;
+
+    move-result-object p0
+
+    const-string v0, "assisted_gps_enabled"
+
+    const/4 v1, 0x0
+
+    invoke-static {p0, v0, v1}, Landroid/provider/Settings$Global;->getInt(Landroid/content/ContentResolver;Ljava/lang/String;I)I
+
+    move-result p0
+
+    const/4 v0, 0x1
+
+    if-ne p0, v0, :cond_0
+
+    move v1, v0
+
+    :cond_0
+    return v1
+.end method
+
+.method public lbsNlpStatusUpdate()V
+    .locals 3
+
+    .line 1622
+    invoke-direct {p0}, Lcom/mediatek/location/lppe/main/LPPeService;->isAssistedGpsEnabled()Z
 
     move-result v0
 
@@ -3090,7 +3138,7 @@
     invoke-direct {p0}, Lcom/mediatek/location/lppe/main/LPPeService;->lbsBatteryForceUpdate()V
 
     .line 1574
-    invoke-direct {p0}, Lcom/mediatek/location/lppe/main/LPPeService;->lbsNlpStatusUpdate()V
+    invoke-virtual {p0}, Lcom/mediatek/location/lppe/main/LPPeService;->lbsNlpStatusUpdate()V
 
     return-void
 .end method
-- 
2.56.0

